### ISSUE TITLE
Debugger stopped working

### ISSUE THREAD

**Issue by pepeu93** [zero-plusplus/vscode-autohotkey-debug#197]
Hello again 🧛 facing a new issue. 
I launched a script then checked it wasn't running, when I look on task manager it was using like 12% CPU but nothing running, this script usually uses a max of 1%, then I killed the process and tried to run again.

Now nothing runs anymore, the exe is launched but the debugger doesn't even read one line of the script.
When I try to click on the AutoHotkey tray icon nothing happens.

I did a > Reload Window, it start working again. This also sucks as you lose ctrl c buff history :(

-Edit-
Looks like this is happening with frequency:
![image](https://user-images.githubusercontent.com/82878129/151583504-b989c406-737b-41ff-a817-2d12b50c5efb.png)
Has been in this state for 2minutes nothing launched yet.

**Comment by zero-plusplus** [comment]
Thanks for the report.

I do not understand the situation very well.
Could you please tell me the specific steps to reproduce the issue?

For example.
1. Debug a script with the following code
```ahk
example := ""
```
2. Verify that it is not running in the task manager

---

Also, did this issue occur in previous versions? I would especially like to know how it was with 1.9.0 and 1.10.0.

**Comment by pepeu93** [comment]
Hi, sorry late reply.

> Also, did this issue occur in previous versions? I would especially like to know how it was with 1.9.0 and 1.10.0.

Yes, I remember this to occur from very old versions.

I don't think it's a problem caused by code, as I have been getting this issue independent of which code the script contains.

I have the same hotkey to start and restart:
![image](https://user-images.githubusercontent.com/82878129/151728849-9a255460-8888-4196-bde0-bcf1bcbc5e41.png)

If you could try it, for example, you have a script running you hit F5 then it will restart the debugger, then there's a chance to happen this issue:
![image](https://user-images.githubusercontent.com/82878129/151729174-1e93b68a-6e12-4152-8781-64c56194b9e3.png)

And then there's a chance to this:
![image](https://user-images.githubusercontent.com/82878129/151729264-685e7dd4-8cc3-4eb3-b534-45bd3e155df2.png)

The script never launch, and the exe continues there until you kill it.

**Comment by zero-plusplus** [comment]
> Hi, sorry for the late reply.

No problem.

---

I am still unable to reproduce this issue.

> You have a script running you hit F5 it will restart the debugger and launch the script again, then there's a chance to happen this issue:

Are you saying that the issue occurs when starting to debug an AutoHotkey script that is running in an unrelated part of the this extension and then restart it?

---

> Failed to connect to an active debugger client.

The above message is generated when the initial communication with the debugger times out.

**Comment by pepeu93** [comment]
> Are you saying that the issue occurs when starting to debug an AutoHotkey script that is running in an unrelated part of the this extension and then restart it?
I didn't understand well, the issue occurs when i launch the debug again and the script is already running not necessarily paused in a breakpoint.

**Comment by zero-plusplus** [comment]
Sorry. I don't know the procedure to generate the error.

> I launch the debugger again and the script is already running,

Do I simply start debugging the script and then press the restart button?

**Comment by zero-plusplus** [comment]
Sorry. I had never had this issue before, so I assumed that there was a special operation.

The issue seems to be caused by the fact that there is no way to handle a timeout of the initial communication with the debugger.

Also, the reason why the script is locked to prevent it from exiting is because it is running in debug mode. Therefore, the debug adapter needs to terminate this.

> It happens randomly I could not find a way to replicate it 100% of the time.

This issue can be reproduced by deliberately causing a timeout, so there is no need to find it.

---

The fixed version will be released within a week. However, it may be delayed due to time constraints.

**Comment by zero-plusplus** [comment]
> Failed to connect to an active debugger client.

The following is what happens when clicking the button displayed in the above message.

* `Abort`: The script terminates. The debugging session will also be terminated accordingly
* `Retry`: The debugging session is started and the script is executed
* `Ignore`: The debug session is terminated normally as in `Abort`

---

To solve this issue there is a way to reproduce it with high probability, or even share the project folder where the issue occurs.
The project folder should be as small as possible, created specifically for this purpose.
It should contain files related to debugging such as `.vscode/launch.json`.

**Comment by pepeu93** [comment]
Also, when you hit "F5" to reload the debugger, it doesn't go to any existing `[OnExit](https://www.autohotkey.com/docs/commands/OnExit.htm)` label, I think it just shutdown "instantly" and reload.

Is possible to go first to any `OnExit` label before reloading, in this case?

**Comment by zero-plusplus** [comment]
When restarted, it sends a command to the debugger to stop the script.

It is probably a specification of the debugger that OnExit does not work.

**Comment by pepeu93** [comment]
Do you mean it's a vscode thing and not the extension, which is causing it to skip the OnExit label? (Sorry I don't speak English well.)

**Comment by pepeu93** [comment]
Im asking because i have a function that runs `onExit` that get the value of all controls in the GUI and save it into a `.ini`, so when i run it again i can load the previous values, but when i reload with `F5` it get skipped.
Any idea in how i could save even at this occasion?

**Comment by zero-plusplus** [comment]
This is probably not possible because AutoHotkey cannot evaluate scripts from the outside like other languages can.

However, there is a workaround. Insert the following code into your script.

```
#If WinActive("ahk_exe Code.exe")
F5::
  OnExitFunc() ; Change this to the function you have registered for your OnExit.
  Send {F5}
  return
#If
```

This would allow you to save the GUI values before restarting debug.

**Comment by pepeu93** [comment]
Good idea! but best to use PostMessage: `PostMessage, 256, 116, 4128769,, A` 
`Send` sometimes doesn't work.
Anyway thanks :)

About the `Failed to connect to an active debugger client.` thing, will be fixed in future updates?

**Comment by zero-plusplus** [comment]
As an advertisement, I have a [my extension](https://marketplace.visualstudio.com/items?itemName=zero-plusplus.vscode-operate-from-autohotkey) to run VSCode commands without using shortcuts.

It's a little cumbersome because you can't just install it and use it.

---

> About the Failed to connect to an active debugger client. thing, will be fixed in future updates?

Since the issue cannot be reproduced, it cannot be fixed.

All the ones I have reproduced here are working fine, so I think it is a different issue.

**Comment by pepeu93** [comment]
> To solve this issue there is a way to reproduce it with high probability, or even share the project folder where the issue occurs.
The project folder should be as small as possible, created specifically for this purpose.
It should contain files related to debugging such as .vscode/launch.json.

I thought you had found a way to reproduce it.

**Comment by zero-plusplus** [comment]
The same error message can be displayed, but the subsequent behavior seems to be different.
In my case, it continues or exits normally no matter which button I click.

In your case, what is the behavior when you click on a button?

**Comment by pepeu93** [comment]
It closes but keep open in the memory, i need to kill it on task manager.

**Comment by zero-plusplus** [comment]
Does it have the same issue no matter which button you click?

Also, have you confirmed that this happens even with minimal code, for example
```ahk
return
```

**Comment by pepeu93** [comment]
I'm doing some tests trying to find a way to replicate it, btw:

I thought this had been fixed but I'm still getting:
```
Session closed for the following reasons: read ECONNRESET
Debugging disconnected. AutoHotkey script is continued.
```

**Comment by zero-plusplus** [comment]
> I thought this had been fixed but I'm still getting:

Under what circumstances did it occur?

> If the msgbox Failed to connect to an active debugger client. appears and i click here:
> instead of clicking in any of the options, then everything stop working, i need to reopen vscode.

I can't fix this issue unless I can actually reproduce the problem.

Can you provide the project folder that contains the launch.json that causes this issue?
It should be minimal code.

For example, the file structure is as follows.
```
- root
  - .vscode
    - launch.json
  - main.ahk
```

For example, main.ahk
```ahk
return
```

**Comment by zero-plusplus** [comment]
> Session closed for the following reasons: read ECONNRESET
Debugging disconnected. AutoHotkey script is continued.

Did you still need to restart VSCode when you saw the above message?

---

This is probably not a sure fix, but we plan to fix the suspicious part so that the above message will not be displayed.

**Comment by pepeu93** [comment]
--------------------------------------------------

> Did you still need to restart VSCode when you saw the above message?

Sometimes, in the actual version I'm getting 80% less:

> Session closed for the following reasons: read ECONNRESET
> Debugging disconnected. AutoHotkey script is continued.





--------------------
> Under what circumstances did it occur?


I have added a MsgBox in the line:
https://github.com/Lexikos/AutoHotkey_L/blob/5d5a569f123f68cbd522d90bf6780f2b3be404af/source/script2.cpp#L12821

```c++
		auto line = g_script.mCurrLine;
		auto mLineNumber = line->mLineNumber;
		auto text = line->mArg->text;

		LPCTSTR msg;
		auto ecode = GetExceptionCode();
		switch (ecode)
		{
			// Having specific messages for the most common exceptions seems worth the added code size.
			// The term "stack overflow" is not used because it is probably less easily understood by
			// the average user, and is not useful as a search term due to stackoverflow.com.
		case EXCEPTION_STACK_OVERFLOW: msg = _T("Function recursion limit exceeded."); break;
		case EXCEPTION_ACCESS_VIOLATION: msg = _T("Invalid memory read/write."); break;
		default: msg = _T("System exception 0x%X."); break;
		}

		TCHAR buf[2000];
		sntprintf(buf, _countof(buf), _T("Exception: \n\n%s \nError Code: %d \n\nLine: %d \n%s"), msg, ecode, mLineNumber, text);
		if (g->ExceptionWarning)
			MessageBox(0, buf, L"", MB_OK | MB_ICONSTOP | MB_SETFOREGROUND | MB_APPLMODAL | MB_TOPMOST);
```

To be able to see when exceptions happen. And I have noted that 100% of the time the debug stops working (ECONNRESET)  I get a MsgBox before, like this:
![image](https://user-images.githubusercontent.com/82878129/153015668-8554a4aa-df05-44e2-8e22-e5f48f899ad3.png)
In the line `1909` had nothing special, just was assigning value to a variable `Text:="8527468-81,61,12,13" `

And 100% of the time it happens the script is currently paused in the line shown in the MessageBox.

> I can't fix this issue unless I can actually reproduce the issue.

 I know it's hard to fix this issue without an example code to you be able to reproduce, but this happens which totally random code, even simple codes.
My script contains 38K lines its complicated to share it, and as said it happens with totally random lines, even lines that are just assigning simple values to variables, but always there's a breakpoint in these lines if you want to take a look at it we could talk in discord.

**Comment by zero-plusplus** [comment]
> Sometimes, in the actual version I'm getting 80% less:

The fact that you don't necessarily need to restart VSCode when ECONNRESET is displayed may mean that the issue is not occurring in only one place.
If the issue is occurring in two or more locations, it may be more difficult to resolve.

> To be able to see when exceptions happen. And I have noted that 100% of the time the debug stops working (ECONNRESET) I get a MsgBox before, like this:

I've seen the `Invalid memory read/error` error message before. I think this happened when I tried to rewrite the value of a particular variable.

As I thought, it seems that the command is sent even though the AutoHotkey debugger is exiting, and it is firing because it is not exiting correctly at that time.

I forgot to tell you that you can check the AutoHotkey debugger command that was executed by setting launch.json to true for the `trace` attribute.

Please use this to check the timing of the issue.

> And 100% of the time it happens the script is currently paused by a **breakpoint** in the line shown in the MessageBox.

When stopping at a breakpoint, a lot of data is retrieved from the AutoHotkey debugger. There is probably a issue somewhere in this timing.

The one that seems most probable is the data inspection process.
This is because I remember that the Invalid error above also occurred when retrieving a property.

> I know it's hard to fix this issue without an example code to you be able to reproduce, but this happens which totally random code, even simple codes.

By `simple codes` do you mean the codes contained in your 38K lines?
I would like to know if it happens on your project or even on a few lines of script.

> If you want to take a look at it we could talk in discord, I can mail you my discord user, I'm up to help with anything you need in the extension, and not just about this current issue 👍 .

Thank you. But I don't plan to use discord, because I want to keep track of all topics related to this extension on GitHub.
Also, it is not practical in terms of time to verify the 38k source code, so I give up.

---

The latest develop branch fixes some of the areas that would cause ECONNRESET. You can clone this project and create a vsix to install the extension by running the following command prompt commands.

1. `npm install`
2. `npm run package`

You can try it if you like.

**Comment by pepeu93** [comment]
> I forgot to tell you that you can check the AutoHotkey debugger command that was executed by setting launch.json to true for the trace attribute.
Please use this to check the timing of the issue.

I have added it, i'll repply with the command when the issue happen.

> By simple codes do you mean the codes contained in your 38K lines?

No, i get it in any script, but it's always on random lines when paused by a breakpoint. But in this actual version im rarely facing this issue.

>The latest develop branch fixes some of the areas that would cause ECONNRESET

I have never worked with vscode extensions or vsix, ill try it.

**Comment by pepeu93** [comment]
I got the ecconreset thing right now with this script:
```
Gui, +LastFound
;DllCall("SetClassLong", "uint", WinExist(), "int", -26
;    , "int", DllCall("GetClassLong", "uint", WinExist(), "int", -26) | 0x20000)


Gui, +HwndhGUI -Caption +E0x02000000 +E0x00080000
Gui, Color, 0
Gui, Show, w640 h480
Return


AnimateWindow(Hwnd, Dir, Speed) {
   WinGetPos,,, W, H, ahk_id %Hwnd%

   W2  := W/10
   If (Dir = "Right")
      W := 0

   Loop, 10 {
      If (Dir = "Right")
         W -= W2
      Else
         W += W
      WinSet,Region, 0-0 W%W2% H%H%, ahk_id %hGUI%
      Sleep, %Speed%
   }
}

F2::
AnimateWindow(hGui, "Left", 100)
Return
```
I was paused by a breakpoint at line 31, `W += W` and it has thrown that MessageBox about Invalid memory read-write in exactly the line 31.

This is what the command outputted in the debug console after im closing the MessageBox:
```
2022/01/11 17:01:09.342 Trace session error
Session closed for the following reasons: read ECONNRESET
2022/01/11 17:01:09.342 Trace sendTerminateEvent
```

**Comment by zero-plusplus** [comment]
> I got the ECONNRESET thing right now with this script:

I was not able to confirm this in my environment. However, I confirmed that ECONNRESET is displayed when I added the following code to AutoExec-Section in [demo.ahk2](https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/master/demo/demo.ahk2) and debugged it.
```
a := 1
b := a += 1
return
```

However, I could not see this problem after the second time.

I don't know if it's a coincidence, but the `+=` operator is used in both examples.

Hmmm, a redesign of the debug adapter may be necessary to solve this issue...

> My apologies but looks like the VSCode console also 'bugged'(?)

I'm having this problem too. It seems that scrolling cancels the selection.

---

Images and videos can be uploaded directly to GitHub by dragging and dropping them.
Please upload as much as possible to Github, as I want to keep all records on GitHub.

**Comment by pepeu93** [comment]
Do you still need me to collect the trace log?
Now I got the error in just these lines:
![image](https://user-images.githubusercontent.com/82878129/153722216-8b3b2d87-ee1a-4b48-a0ec-130b4048ff27.png)

The object was written wrong:
```ahk
Object := {1:X: 30, Tab_Name: "test"}
for index, tab in object
Gui, Add, Button, w300 h300, % Tab.Tab_Name
```
```
: ==> Missing "key:" in object literal.
AutoHotkey closed for the following exit code: 2
```

**Comment by zero-plusplus** [comment]
> Do you still need me to collect the trace log?

Due to bugs in VSCode, it is not practical to share the trace log, so it is not necessary.

---

I have checked your example, and the debug adapter exits successfully.

I would like to check, have you cleared all breakpoints (except one) and watch expression?

This issue can be caused by them.

**Comment by pepeu93** [comment]
I was getting exceptions constantly in this script, even after calling a >Reload.

I did what you said, I cleared all breakpoints in all scripts and also the watch window.
Then I did a >Reload, set a breakpoint in the same line that was throwing the exception, launch the script, the except was gone.

**Comment by zero-plusplus** [comment]
Where did you call Reload in your example?

**Comment by zero-plusplus** [comment]
Can you use Windows SandBox available in Windows 10 Pro?

If this is available, you can clone the master branch of this repository and use the following command prompt commands to build a test environment.

1. `npm install`
2. `npm run test:sandbox`

By using this test environment, you can find out if it is an environment-dependent bug or not.

**Comment by zero-plusplus** [comment]
What is the reason for closing this issue?

**Comment by pepeu93** [comment]
@zero-plusplus sorry I didn't note I had closed, should I reopen it as the issue still persist?

Does somebody else report this issue? any news regarding this issue?

**Comment by zero-plusplus** [comment]
> sorry I didn't note I had closed, should I reopen it as the issue still persist?

OK, I reopened this issue.

> Any news regarding this issue? does somebody else report it?

Unfortunately, there is no new information on this issue.

> I have a class that handles many window messages using OnMessage and I noted that when I comment the line that enables it OnMessage.New() this issue of Session closed for the following reasons: read ECONNRESET Debugging disconnected. happen way less.

This is interesting. I think the probability is low, but if it changes whether or not it is more likely to occur with a particular script, then it could be a bug in the AutoHotkey debugger itself. In that case, we may see the same bug in SciTE4AutoHotkey.

**Comment by zero-plusplus** [comment]
I have confirmed that ECONNRESET occurs with the following code.

The condition is "Execute this code, hit step-in and exit the script. In my environment, it occurs about once every 10 times.

I have seen it occur in both v1 and v2.
```ahk
; @Debug-Breakpoint
a := 1
a := 1
```


However, it seems that the debugger is terminated correctly because this error does not cause a bug that prevents watch expressions from being registered. Therefore, there was no need to reload the window.

The error message is the same, but the cause seems to be different from this issue.

**Comment by pepeu93** [comment]
It also happens here, sometimes the debug continues working when ECONNRESET occurs, and sometimes it stops completely, then I need to reload.

Some other things I already tried were:
> The required version of VSCode has been raised from [1.45.0](https://code.visualstudio.com/updates/v1_45) to [1.49.0](https://code.visualstudio.com/updates/v1_49).

- Downgrading vscode to 1.49 as it's the latest supported.
- Uninstalling everything and installing it again, I also test on the other two computers I have, and I got the issue in them all (one of them I formated has few weeks).

- `ECONNRESET` happens in different scripts, even the simplest ones with just a very few lines like your example above, but its completely random.

**Comment by zero-plusplus** [comment]
Thanks for testing.
Can you confirm that this also occurs in v1.0.0, which was a simpler implementation?

If you cannot confirm the issue with v1.0.0, please upgrade to see which version is causing the issue.

If you want to check efficiently, check the middle of the versions you want to test.
For example, test from v1.0.0->v1.6.0, and if the issue is confirmed, check v1.3.0, and if not, check v1.9.0.

Also, make sure your vscode version is the latest when testing.

**Comment by pepeu93** [comment]
1.9.0 the issue happens
1.7.0 dont

But i have tested just few hours, should i test maybe a week?

**Comment by zero-plusplus** [comment]
Please upload all gifs and other images to be shared to Github instead of using imgur.
Images previously uploaded to imgur will no longer be visible and cannot be checked later.

---

> --Edit3--
Looks like the issue started on version 1.3.7, just hovering a variable and occurs ECONNRESET. https://i.imgur.com/R40bTCW.gif

The "Invalid memory read/write" error was fixed in previous versions, but may not have been fixed as of 1.3.7.

---

I am a little confused.
There are multiple causes of ECONNRESET errors, so you need to be consistent in how checks are done.

In other words, you need to check each version of the same source code for the same issue that occurs at the same location.

Also, the errors to be checked should be limited to fatal ones that require reloading.

Does the gif you provided meet the above requirements?

**Comment by pepeu93** [comment]
> you need to check each version of the same source code for the same issue that occurs at the same location.

This is how I have checked it.

> the errors to be checked should be limited to fatal ones that require reloading.

Whenever the ECONNRESET happens in my tests I need to reload.

> Does the gif you provided meet the above requirements?

Yes. I have checked each time it throw ECONNRESET if the debugger still was working and it wasnt.

**Comment by zero-plusplus** [comment]
Thanks.

The file you are checking has about 30,000 lines. Can you cut it down to, for example, function units and reduce the size of the file to be tested?

If you can finally reduce it to a dozen or so lines, I can join in the identification work.

---

> Another thing I have noted is if I press ctrl+v and past anything in the editor it doesn't error, but as soon I type something it does, this doesn't suggest anything?


Since [v1.6.0](https://github.com/zero-plusplus/vscode-autohotkey-debug/blob/master/CHANGELOG.md#160---2020-10-04) or later, IntelliSense is implemented, so errors may occur when inputting data. It should not occur in versions less than that.

**Comment by pepeu93** [comment]
Hello, I spent the entire day trying to reproduce the error with fewer lines, I was able to reproduce it with a script containing 2500 lines where 2100 lines are just the GDI ahk lib.
![Code_WyzCy7bt04](https://user-images.githubusercontent.com/82878129/168447840-05ad8c69-2506-4943-a8f8-29af14163fdf.gif)

This is the script i used:
https://pastebin.com/h2fNSqZe

Steps to reproduce:
launch the script > hover the `obj[1]` at line 17 > hover the string `OnMessage` at line 9
ECONNRESET occur.

I show in the gif that after launching the script again nothing work anymore.

**Comment by zero-plusplus** [comment]
If you want to share the source code, please upload it to GitHub. You should be able to upload it by drag-and-drop, just like you did with the video.

---

I tried but could not reproduce it in my environment.

From the video, it seems that the project structure such as workspace is different.

Please upload a zip file of the project folder with the minimum configuration including launch.json etc. that can reproduce the issue.

**Comment by pepeu93** [comment]
I confirmed that the issue is being caused **because of modifications I did on the source of the AutoHotkey**, I'm closing the issue as the extension is not causing it.

**Comment by zero-plusplus** [comment]
OK.
Never mind, thanks for the confirmation work.

Please let me know if there are any more issues.
